File: 2018/go/d09/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strings"
)

type Marble struct {
	left *Marble
	right *Marble
	value int
}

func (m Marble) String() string {
	res := strings.Builder{}
	next := &m
	res.WriteString(fmt.Sprint(next.value))
	next = next.right
	for next != nil &&  *next != m {
	//	fmt.Printf("%p %v", next.right, &m)
		res.WriteString(" ")
		res.WriteString(fmt.Sprint(next.value))
		next = next.right
	}
	return res.String()
}

func (m *Marble) AddLeft(n *Marble) {
	m.left, n.left = n, m.left
	n.right = m
	n.left.right = n
}

func (m *Marble) AddRight(n *Marble) {
	m.right.AddLeft(n)
}

func (m *Marble) Remove() {
	m.right.left = m.left
	m.left.right = m.right
}

func main() {
	f, err := os.Open("input.txt")
	if err != nil {
		panic("Could not read input")
	}
	b, err := ioutil.ReadAll(f)
	if err != nil {
		panic(err)
	}
	var np, max int
	fmt.Sscanf(string(b), "%d players; last marble is worth %d points", &np, &max)
	fmt.Println(np, max)

	scores := make([]int, np)

	current := &Marble{value: 0}
	current.left = current
	current.right = current
	for i:=1; i<max; i++ {
		if i % 23 != 0 {
			new := &Marble{value: i}
			current.right.AddRight(new)
			current = new
		} else {
			player := (i-1) % np
			scores[player] += i
			for i:=0; i<7; i++ {
				current = current.left
			}
			scores[player] += current.value
			current.Remove()
			current = current.right
		}
	}
	winner := 0
	for _, val := range scores {
		if val > winner {
			winner = val
		}
	}
	fmt.Println(winner)

}

File: 2018/rust/d09/src/main.rs
use std::fs;
use std::collections::LinkedList;

fn main() {
    let input = fs::read_to_string("input").expect("could not read file");
    println!("Solution 1: {}", solve1(&input));
    println!("Solution 2: {}", solve2(&input));
}

fn play1(nplayers: usize, last: usize)  -> usize{
    let mut marbles = LinkedList::from([0]);
    let mut scores = vec![0; nplayers];
    for i in 1..=last {
        if i % 23 == 0 {
            let player = i % scores.len();
            scores[player] += i;
            for _ in 0..6 {
                let removed = marbles.pop_back().unwrap();
                marbles.push_front(removed);
            }
            scores[player] += marbles.pop_back().unwrap();

        } else {
            for _ in 0..2 {
                let removed = marbles.pop_front().unwrap();
                marbles.push_back(removed);
            }
            marbles.push_front(i);
        }
    }
    scores.iter().max().unwrap().to_owned()
    
}

fn parse(input: &str) -> (usize, usize) {
    let tokens: Vec<&str> = input.split(' ').collect();
    let nplayers = tokens[0].parse().unwrap();
    let last = tokens[6].parse().unwrap();
    (nplayers, last)
}

fn solve1(input: &str) -> usize{
    let (nplayers, last) = parse(input);
    play1(nplayers, last)

}
fn solve2(input: &str) -> usize{
    let (nplayers, last) = parse(input);
    play1(nplayers, last * 100)
}

#[test]
fn test_example(){
    assert_eq!(solve1("10 players; last marble is worth 1618 points"), 8317);
    assert_eq!(solve1("13 players; last marble is worth 7999 points"), 146373);
    assert_eq!(solve1("17 players; last marble is worth 1104 points"), 2764);
    assert_eq!(solve1("21 players; last marble is worth 6111 points"), 54718);
    assert_eq!(solve1("30 players; last marble is worth 5807 points"), 37305);
}
